### ISSUE TITLE
String literal to memory conversion for conditional operator

### ISSUE THREAD

**Issue by dddejan** [SRI-CSL/solidity#89]
```solidity
pragma solidity >=0.5.0;
contract A {
  function sign(int x) public view returns (string memory result) {
    result = x > 0 ? "pos" : "neg";
  }
}
```

Gives 

```
solc-verify.py Issue.sol --output .
Error while running verifier, details:
Parsing ./Issue.sol.bpl
./Issue.sol.bpl(22,1): Error: mismatched types in assignment command (cannot assign int_arr_type to int_arr_ptr)
1 type checking errors detected in ./Issue.sol.bpl
```

**Comment by hajduakos** [comment]
This is tricky. Having simply `result = "str";` works, because the LHS is `string memory` and RHS is `string literal`, which we treat as storage and we make the appropriate conversion. However, if there is a conditional expression `x > 0 ? "pos" : "neg"` where both the true and false branch is a `string literal`, the type of the whole expression is not `string literal`, but `string memory` so we do not make the conversion anymore. One way to solve this is to introduce temp variables for both branches with the type of the whole expression and make an assignment. This would do the conversions and then the temp vars could be used in the Boogie conditional. However, we have to make sure that this only happens if it is really needed (e.g., this case). Otherwise, conditional expressions could not be used in specs (as they would always introduce side effects).

**Comment by dddejan** [comment]
Yes. I started treating string literals as storage as it seems more natural. I would also treat the conditional result as storage (i.e. just an ITE over the two values). I think this is sound.

**Comment by hajduakos** [comment]
I am not sure how that would work out. First, it is the compiler that marks the result as string memory. To override this, we would have to insert checks all over the code where we ask for a type of an expression. Keeping it inside the conditional is more localized. Second, there can be cases where the true branch is a string literal and the false branch is a string memory.

Thus, I would prefer doing the necessary conversion/allocation within the conditional.

**Comment by hajduakos** [comment]
See also #92

**Comment by hajduakos** [comment]
Fixed in cc77d81592c9d077e171cd2e41e17d710c63fa82, but not yet merged

**Comment by hajduakos** [comment]
@dddejan #94 was merged, please check and close
